src/main/scala/firrtl/options/StageAnnotations.scala

```diff
@@ -3,40 +3,89 @@
 package firrtl.options
 
 import firrtl.AnnotationSeq
-import firrtl.annotations.NoTargetAnnotation
+import firrtl.annotations.{Annotation, NoTargetAnnotation}
 
 import scopt.OptionParser
 
-sealed trait StageOption extends HasScoptOptions
+sealed trait StageOption { this: Annotation => }
+
+/** An annotation that should not be serialized automatically [[phases.WriteOutputAnnotations WriteOutputAnnotations]].
+  * This usually means that this is an annotation that is used only internally to a [[Stage]].
+  */
+trait Unserializable { this: Annotation => }
```

Comment:
Should there be a class of Unserializables that error if you try to serialize them, instead of silently not serializing?

Reply:
I actually think that all `Unserializable`s are the same. However, you want to be able to change the behavior of this in certain scenarios. Two scenarios come to mind:

1. User writes a `Stage` and the automatic `WriteOutputAnnotations` runs. 
2. User wants to run `WriteOutputAnnotations` after every `Phase` for debugging reasons.

In the former case, an `Unserializable` is a "stage local" annotation that was supposed to be removed, deleted, or just not supposed to exist at the end of the `Stage`. In the latter case, an `Unserializable` may exist, but this is okay and should be ignored. 

It initially seemed like this means `WriteOutputAnnotations` should be parameterized. However, I now think that `WriteOutputAnnotations` should ignore `Unserializable`. I need to add additional `Checks` that run at the end of a `Stage` to verify that all `Unserializable` were removed.

The assumption here is that `Unserializable`s are not supposed to exit a `Stage`.

What do you think? Thoughts @chick?

Reply:
Thanks for the thoughts below @chick. Certain `Unserializable` annotations do currently need to exit `ChiselStage` and these therefore can't be entirely purged.

I think things should likely then be left the way they are for now.

I do think that this hints at a larger problem: there needs to be some way of passing information from earlier in the pipeline to later in the pipeline. This can likely be done either by splitting the `AnnotationSeq` as it goes along or with a modification to the existing `DeletedAnnotation`. 

`DeletedAnnotation` doesn't really communicate what you need here. You want to have some history about what was used to create a specific `Annotation`. There could be an optional `evidence` member added to an `Annotation` that can include information about what was used to create that `Annotation`. 

E.g., we currently have the following:

  - `ChiselGeneratorAnnotation(gen: () => RawModule)`
  - `ChiselCircuitAnnotation(circuit: chisel3.internal.firrtl.Circuit)`
  - `FirrtlCircuitAnnotation(circuit: firrtl.ir.Circuit)`
  - `EmittedAnnotation[T <: EmittedComponent]{ def value: T }`

The end result of a run of `ChiselStage` -> `FirrtlStage` would then be one or more `EmittedAnnotation` that includes evidence about how it got to that point. The evidence would be excluded from the serialization or could be converted to `DeletedAnnotation`s.

Alternatively, the notion of splitting an `AnnotationSeq` could be resolved with the `DependencyAPI`. 

As a further alternative, testers and treadle could use the constituent `Phase`s to construct a different `Stage` if that made sense.